### ISSUE TITLE
nDPI and big endianess

### ISSUE THREAD

**Issue by utoni** [ntop/nDPI#1312]
Although big endian architectures may not be very popular these days, nDPI has some serious issues dealing with them.

I've compiled nDPI for MIPS and ran `tests/do.sh` to verify that there are some.
Most of the issues are related to the recent ahocorasick library update in 65678dbeeabcf51e02ba3cc9ffbd36324a92a971. But there are also at least half a dozen endian issues in protocol dissectors.

I've created a guide (as Gist) how to reproduce those issues:
[HowTo-nDPI-MIPS.md](https://gist.github.com/lnslbrty/918abd33e7e99234d6c19b2800db8b53)

I would like to dig deeper in the ahocorasick library, but I was not able to find any upstream sources for the library which is used in nDPI. Any hints?
To prevent such issues in the future, a CI job building and testing for a big endian architecture may help.

Cheers!

**Comment by utoni** [comment]
Some failing pcaps, caused by host patterns not found in ac automata:
```
qemu-mips -L /home/toni/x-tools/sysroot-mips-linux-gnu /home/toni/x-tools/sysroot-mips-linux-gnu/usr/bin/ndpiReader -p ../example/protos.txt -c ../example/categories.txt -r ../example/risky_domains.txt -j ../example/ja3_fingerprints.csv -S ../example/sha1_fingerprints.csv -q -t -i pcap/443-curl.pcap -w /tmp/reader.out -v 2 [old vs new]
1c1
< ntop	109	73982	1
---
> TLS	109	73982	1
8c8
< 	1	TCP 192.168.1.13:55523 <-> 178.62.197.130:443 [proto: 91.26/TLS.ntop][cat: Network/14][51 pkts/4260 bytes <-> 58 pkts/69722 bytes][Goodput ratio: 22/94][1.10 sec][ALPN: h2;http/1.1][bytes ratio: -0.885 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 25/19 784/784 122/114][Pkt Len c2s/s2c min/avg/max/stddev: 54/66 84/1202 583/1506 74/562][TLSv1.2][Client: www.ntop.org][JA3C: 2a26b1a62e40d25d4de3babc9d532f30][ServerNames: www.ntop.org][JA3S: ae53107a2e47ea20c72ac44821a728bf][Issuer: C=US, O=Let's Encrypt, CN=Let's Encrypt Authority X3][Subject: CN=www.ntop.org][Certificate SHA-1: DB:A7:E4:3E:6D:BB:21:AB:68:47:35:E8:0B:8F:15:DF:DB:C7:C9:6F][Firefox][Validity: 2019-12-17 01:17:28 - 2020-03-16 01:17:28][Cipher: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256][Plen Bins: 3,13,1,1,1,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,1,0,73,0,0]
---
> 	1	TCP 192.168.1.13:55523 <-> 178.62.197.130:443 [proto: 91/TLS][cat: Web/5][51 pkts/4260 bytes <-> 58 pkts/69722 bytes][Goodput ratio: 22/94][1.10 sec][ALPN: h2;http/1.1][bytes ratio: -0.885 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 25/19 784/784 122/114][Pkt Len c2s/s2c min/avg/max/stddev: 54/66 84/1202 583/1506 74/562][TLSv1.2][Client: www.ntop.org][JA3C: 2a26b1a62e40d25d4de3babc9d532f30][ServerNames: www.ntop.org][JA3S: ae53107a2e47ea20c72ac44821a728bf][Issuer: C=US, O=Let's Encrypt, CN=Let's Encrypt Authority X3][Subject: CN=www.ntop.org][Certificate SHA-1: DB:A7:E4:3E:6D:BB:21:AB:68:47:35:E8:0B:8F:15:DF:DB:C7:C9:6F][Firefox][Validity: 2019-12-17 01:17:28 - 2020-03-16 01:17:28][Cipher: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256][Plen Bins: 3,13,1,1,1,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,1,0,73,0,0]
```

```
qemu-mips -L /home/toni/x-tools/sysroot-mips-linux-gnu /home/toni/x-tools/sysroot-mips-linux-gnu/usr/bin/ndpiReader -p ../example/protos.txt -c ../example/categories.txt -r ../example/risky_domains.txt -j ../example/ja3_fingerprints.csv -S ../example/sha1_fingerprints.csv -q -t -i pcap/4in6tunnel.pcap -w /tmp/reader.out -v 2 [old vs new]
1c1
< Microsoft	4	2188	1
---
> TLS	4	2188	1
8c8
< 	1	TCP 192.168.0.1:64455 <-> 10.10.10.1:443 [proto: 91.212/TLS.Microsoft][cat: Web/5][2 pkts/520 bytes <-> 2 pkts/1668 bytes][Goodput ratio: 43/82][< 1 sec][ALPN: h2;http/1.1][TLSv1.2][Client: www.bing.com][JA3C: 9e10692f1b7f78228b2d4e424db3a98c][Firefox][Plen Bins: 0,0,0,0,0,0,0,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,50,0,0,0,0,0]
---
> 	1	TCP 192.168.0.1:64455 <-> 10.10.10.1:443 [proto: 91/TLS][cat: Web/5][2 pkts/520 bytes <-> 2 pkts/1668 bytes][Goodput ratio: 43/82][< 1 sec][ALPN: h2;http/1.1][TLSv1.2][Client: www.bing.com][JA3C: 9e10692f1b7f78228b2d4e424db3a98c][Firefox][Plen Bins: 0,0,0,0,0,0,0,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,50,0,0,0,0,0]
```

```
qemu-mips -L /home/toni/x-tools/sysroot-mips-linux-gnu /home/toni/x-tools/sysroot-mips-linux-gnu/usr/bin/ndpiReader -p ../example/protos.txt -c ../example/categories.txt -r ../example/risky_domains.txt -j ../example/ja3_fingerprints.csv -S ../example/sha1_fingerprints.csv -q -t -i pcap/anydesk-2.pcap -w /tmp/reader.out -v 2 [old vs new]
1c1,2
< AnyDesk	2083	346113	4
---
> DNS	4	392	2
> AnyDesk	2079	345721	2
5,6c6,7
< 	1	 192.168.1.187            	 1      
< 	2	 192.168.1.178            	 1      
---
> 	1	 192.168.1.178            	 1      
> 	2	 192.168.1.187            	 1      
11,12c12,13
< 	3	UDP 192.168.1.187:55376 <-> 192.168.1.1:53 [proto: 5.252/DNS.AnyDesk][cat: RemoteAccess/12][1 pkts/90 bytes <-> 1 pkts/106 bytes][Goodput ratio: 53/60][0.01 sec][Host: relay-9b6827f2.net.anydesk.com][138.199.36.115][PLAIN TEXT (anydesk)][Plen Bins: 0,50,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
< 	4	UDP 192.168.1.187:59511 <-> 192.168.1.1:53 [proto: 5.252/DNS.AnyDesk][cat: RemoteAccess/12][1 pkts/90 bytes <-> 1 pkts/106 bytes][Goodput ratio: 53/60][0.01 sec][Host: relay-3185a847.net.anydesk.com][37.61.223.15][PLAIN TEXT (anydesk)][Plen Bins: 0,50,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
---
> 	3	UDP 192.168.1.187:55376 <-> 192.168.1.1:53 [proto: 5/DNS][cat: Network/14][1 pkts/90 bytes <-> 1 pkts/106 bytes][Goodput ratio: 53/60][0.01 sec][Host: relay-9b6827f2.net.anydesk.com][138.199.36.115][PLAIN TEXT (anydesk)][Plen Bins: 0,50,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
> 	4	UDP 192.168.1.187:59511 <-> 192.168.1.1:53 [proto: 5/DNS][cat: Network/14][1 pkts/90 bytes <-> 1 pkts/106 bytes][Goodput ratio: 53/60][0.01 sec][Host: relay-3185a847.net.anydesk.com][37.61.223.15][PLAIN TEXT (anydesk)][Plen Bins: 0,50,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
```

**Comment by aouinizied** [comment]
@lnslbrty I'm working on CI additional jobs (build and test). Especially for specific platforms such as ARM64 etc.
I found a compilation issue that was not triggered using current cross arm-linux-gnueabihf.

I'm using ubuntu-core:arm64  

ndpi_main.c:7120:26: error: expected expression before ')' token
   return(NDPI_API_VERSION);
                          ^
ndpi_main.c:7121:1: error: control reaches end of non-void function [-Werror=return-type]
 }

Do you have any clue about this?

Zied

**Comment by lucaderi** [comment]
I have spent time a year ago on this, and the fix was not quick. I will check again when I have time

**Comment by viniciussn** [comment]
I've compiled nDPI for a MIPS architecture device and some tests failed, e.g. youtube_quic.pcap:

```
Using nDPI (4.1.0-3381-8e512ff5) [1 thread(s)]
Using libgcrypt version 1.9.3
Reading packets from pcap file youtube_quic.pcap...
Running thread 0...

nDPI Memory statistics:
	nDPI Memory (once):      211.51 KB    
	Flow Memory (per flow):  1.01 KB      
	Actual Memory:           1.44 MB      
	Peak Memory:             1.44 MB      
	Setup Time:              248 msec
	Packet Processing Time:  21 msec

Traffic statistics:
	Ethernet bytes:        198575        (includes ethernet CRC/IFC/trailer)
	Discarded bytes:       0            
	IP packets:            289           of 289 packets total
	IP bytes:              191639        (avg pkt size 663 bytes)
	Unique flows:          3            
	TCP Packets:           0            
	UDP Packets:           289          
	VLAN Packets:          0            
	MPLS Packets:          0            
	PPPoE Packets:         0            
	Fragmented Packets:    0            
	Max Packet size:       1358         
	Packet Len < 64:       87           
	Packet Len 64-128:     9            
	Packet Len 128-256:    49           
	Packet Len 256-1024:   33           
	Packet Len 1024-1500:  111          
	Packet Len > 1500:     0            
	nDPI throughput:       13.58 K pps / 71.20 Mb/sec
	Analysis begin:        13/Mar/2017 00:10:23
	Analysis end:          13/Mar/2017 00:10:26
	Traffic throughput:    85.11 pps / 456.90 Kb/sec
	Traffic duration:      3.395 sec
	Guessed flow protos:   0            
	DPI Packets (UDP):     3             (1.00 pkts/flow)


Detected protocols:
	Google               packets: 289           bytes: 191639        flows: 3            


Protocol statistics:
	Tracker/Ads                 191639 bytes
```

Taking a look at the source code, I've found a few places assuming that the host is little-endian, for instance: https://github.com/ntop/nDPI/blob/8e512ff50331e2750d13b3d983f3b3216134ad18/src/lib/protocols/quic.c#L238-L243

Later I've found this issue about problems running nDPI in big-endian architectures (also MIPS in my case). 

Is anyone still working on this issue?
If not, can someone guide the other locations that will need adjustment?

**Comment by IvanNardi** [comment]
AFAIK, nobody is actively working on this issue; any help will be really appreciated!
Looking at the CI tests results, it seems there are multiple, independent, issues on big-endian architectures. The biggest one is about the ahocorasick library: string matching algorithm doesn't always return the expected results.
You are right about the QUIC code: a fix will be quickly pushed (CHLO decoding should be correct, but no SNI matching, though)
Have you found other issues looking at the code?

**Comment by IvanNardi** [comment]
Even after dc60cd0, there are still some issues (see https://github.com/ntop/nDPI/runs/4805759337?check_suite_focus=true)

**Comment by IvanNardi** [comment]
Another step in the right direction in https://github.com/ntop/nDPI/commit/0c70411b1b093279f3d7c09b2b57b491911df84c
Remaining issues: https://github.com/ntop/nDPI/runs/4951153935?check_suite_focus=true (these issues seem to be related to `ndpiReader` application, not `libndpi.so` itself)

**Comment by utoni** [comment]
Fixed with PR #1482.
